python3Packages.hy: 1.0a3 -> 1.0a4; hy: replace it with python3Packages.hy; hyrule: init at 0.1

thiagokokada opened the pull request:
###### Description of changes

Fixes: https://github.com/NixOS/nixpkgs/issues/171428

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

thiagokokada committed:
python3Packages.hy: 1.0a3 -> 1.0a4

thiagokokada committed:
hy: replace it with python3Packages.hy

Add `hyDefinedPythonPackages` parameter to allow backwards compatibility
with interpreters/hy.

Fixes https://github.com/NixOS/nixpkgs/issues/171428.

thiagokokada commented:
CC @fabaff @nixy @Mazurel

fabaff reviewed (commented):
Looks like that I added a "duplicate" at some point. Thanks for cleaning it up.

thiagokokada commented:
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
  </ul>
</details>

thiagokokada commented:
Huh... I think the GitHub source doesn't have the version attribute set:

```
Hy unknown using CPython(main) 3.9.12 on Linux
=> (+ 1 1)
2
=>
```

Compares this to `interpreters/hy`:

```
hy 1.0a1 using CPython(main) 3.9.12 on Linux
=>
```

Let me see if I can patch this.

thiagokokada committed:
python3Packages.hy: fix version

thiagokokada commented:
> Let me see if I can patch this.

Done. I also added a test to ensure that the version is being set correctly.

thiagokokada committed:
python3Packages.hy: add passthru.withPackages attribute

thiagokokada commented:
@ofborg eval

thiagokokada commented:
With commit [24c9621](https://github.com/NixOS/nixpkgs/pull/171432/commits/24c96215cc7902dc72ea712ac8e4c64b545bf126) this works:

```
Welcome to Nix 2.4. Type :? for help.

nix-repl> :lf ./
Added 16 variables.

nix-repl> legacyPackages.x86_64-linux.hy.withPackages (ps: with ps; [ pytest ])
«derivation /nix/store/748nk43sfswh1i8d9jjl0rb647j7hx0f-python3.9-hy-1.0a4.drv»
```

In `master`:

```
❯ nix repl
Welcome to Nix 2.4. Type :? for help.

nix-repl> :lf ./
Added 16 variables.

nix-repl> legacyPackages.x86_64-linux.hy.withPackages (ps: with ps; [ pytest ])
«derivation /nix/store/62b9124kh079v3jr1sngh7hglyz6cfrw-hy-1.0a1.drv»
```

thiagokokada commented:
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
  </ul>
</details>

thiagokokada committed:
python3Packages.hy: update meta

thiagokokada committed:
hylure: init at 0.1

thiagokokada commented:
I added `hyrule` too, since on version 1.0a4 most of Hy stdlib was moved to this library.

WolfangAukang commented:
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
    <li>python310Packages.hyrule</li>
    <li>python39Packages.hyrule</li>
  </ul>
</details>

WolfangAukang commented on pkgs/development/python-modules/hy/default.nix:
```diff
@@ -4,31 +4,43 @@
   pname = "hy";
-  version = "1.0a3";
+  version = "1.0a4";
 
```
```suggestion
format = "setuptools";

```

A minor nitpick, but would be helpful for maintenance.

WolfangAukang commented on pkgs/development/python-modules/hyrule/default.nix:
```diff
@@ -0,0 +1,42 @@
+buildPythonPackage rec {
+  pname = "hyrule";
+  version = "0.1";
+
```
```suggestion
format = "setuptools";

```

Same as from above

AtilaSaraiva reviewed (approved):
Result of `nixpkgs-review pr 171432` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>hy (python39Packages.hy)</li>
    <li>python310Packages.hy</li>
    <li>python310Packages.hyrule</li>
    <li>python39Packages.hyrule</li>
  </ul>
</details>

AtilaSaraiva commented:
I don't understand much of what hy does but it makes sense to do this change.

WolfangAukang reviewed (approved):
LGTM
